tensorflow/python/keras/benchmarks/keras_examples_benchmarks/mnist_conv_custom_training_benchmark_test.py

```diff
@@ -0,0 +1,320 @@
+
+    Returns:
+      Loss value.
+    """
+    train_x, train_y = inputs
+    with tf.GradientTape() as tape:
+      predictions = model(train_x, training=True)
+      loss = self.compute_loss(train_y, predictions, loss_fn, batch_size)
+    grads = tape.gradient(loss, model.trainable_weights)
+    optimizer.apply_gradients(zip(grads, model.trainable_weights))
+    return loss
+
+  @tf.function
+  def distributed_train_step(self,
+                             batch_dataset,
+                             model,
+                             loss_fn,
+                             optimizer,
+                             batch_size,
+                             distribution_strategy):
+    """Train step in distribution strategy setting.
+
+    Arguments:
+      batch_dataset: `tf.data`.
+      model: See `model` in `train_function()` method.
+      loss_fn: See `loss_fn` in `train_function()` method.
+      optimizer: See `optimizer` in `train_function()` method.
+      batch_size: See `batch_size` in `train_function()` method.
+      distribution_strategy: See `distribution_strategy`
+        in `train_function()` method.
+
+    Returns:
+      Sum of per_replica_losses.
+    """
+    per_replica_losses = distribution_strategy.run(
+        self.train_step, args=(
+            batch_dataset, model, loss_fn, optimizer, batch_size,))
+    return distribution_strategy.reduce(
+        tf.distribute.ReduceOp.SUM, per_replica_losses, axis=None)
+
+  def train_function(self,
+                     model,
+                     train_dataset,
+                     loss_fn,
+                     optimizer,
+                     epochs=2,
+                     distribution_strategy=None,
+                     batch_size=256):
+    """Train model in custom training loop and return average
+    train_step_time.
+
+    Arguments:
+      model: Model function to be benchmarked.
+      train_dataset: `tf.data` dataset. Should return a tuple
+        of either (inputs, targets) or (inputs, targets, sample_weights).
+      loss_fn: `tf.keras.losses.Loss` instance.
+      optimizer: `tf.keras.optimizers` instance.
+      epochs: Integer. Number of epochs to train the model.
+        If unspecified, `epochs` will default to 2.
+      distribution_strategy: Distribution strategies. It could be
+      `multi_worker_mirrored`, `one_device`, `mirrored`. If unspecified,
+      `distribution_strategy` will default to 'off'. Note that, `TPU`
+      and `parameter_server` are not supported yet.
+      batch_size: Integer. Number of samples per gradient update.
+      If unspecified, `batch_size` will default to 32.
+
+    Returns:
+      Average train_step_time.
+    """
+    train_step_time_list = []
+    timer = timeit.default_timer
+
+    total_loss = 0.0
+    num_batches = 0
+    for epoch in range(epochs):
+      # Iterate over the batches of the dataset.
+      for batch_dataset in train_dataset:
+
+        start_time = timer()
```

Comment:
I think the `start_time` and `end_time` for the train_step should warp the train_step exactly, for a precise train_step time measurement.

Reply:
But if so, for the `train_step` and `distributed_train_step`, it will write some duplicate codes for recording the time. In addition, I checked the results we ran so far the train_step_time usually is `0.00xx`, I think it's enough for us.

Reply:
SGTM.